### PR TITLE
[dynamo] Restore autocast state between tests in TestCase

### LINKED ISSUE #195487: [dynamo] torch._dynamo.test_case.TestCase does not restore autocast state between tests
### Summary

`torch._dynamo.test_case.TestCase` restores grad mode between tests but not autocast state. All four pieces of autocast state -- the nesting counter, enabled, dtype and cache_enabled -- survive into the next test.

This is not hypothetical: in #194754 a test leaked the nesting counter to -2, a genuinely failing test in a later position was masked by `--reruns=2`, and the result reached CI as a green checkmark. The leak was fixed at its source there, but nothing stops the next one.

`tearDown` already handles exactly this shape of problem for grad mode:

https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/test_case.py#L104-L107

Autocast has no equivalent. Adding one would also put the diagnostic where the state is actually lost -- the warning fires on the test that leaked -- instead of letting it surface several tests later as an unrelated dtype mismatch.

### Notes

There is no getter for the autocast nesting counter from Python, only `torch.autocast_increment_nesting()` / `torch.autocast_decrement_nesting()`, so snapshotting requires a matched increment/decrement pair. Exposing a real getter would be cleaner if we want to avoid that.

### Prototype

I prototyped this and it works: restoring the nesting counter in `tearDown` is on its own enough to make the full `test_ctx_manager.py` pass with #194754's new test written against `nn.Linear`, and it produced zero additional warnings across all of `test_misc.py`, so the blast radius looks small.

Note that clearing the autocast cache in `tearDown` is *not* a substitute -- the counter is still negative during the next test, so a bf16 -> fp16 transition inside that test re-pollutes the cache.

cc @mruberry @mcarilli @ptrblck @leslie-fang-intel @jgong5 @chauhang @penguinwu @voznesenskym @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @jataylo @azahed98 @anijain2305 @williamwen42 @jansel

This issue was written with assistance from an AI coding assistant; the investigation and prototype above were reviewed by me.

### PR DESCRIPTION
**Pull request by aperson30**
Fixes #195487.

`torch._dynamo.test_case.TestCase.setUp`/`tearDown` already snapshots `torch.is_grad_enabled()` and restores it (with a warning) in `tearDown` if a test left it changed, but has no equivalent for autocast. All four pieces of autocast state -- enabled, dtype, cache_enabled, and the nesting counter -- survive past the end of a test into whichever test runs next. A test that leaks this state (e.g. by calling `torch.autocast_increment_nesting()` without a matching decrement, or by using the raw setters instead of the `with torch.autocast(...)` context manager) pollutes every subsequent test in the process, which can surface many tests later as an unrelated dtype mismatch instead of on the test that actually caused it.

The fix snapshots/restores autocast's four pieces of state -- enabled and dtype per device (looping over all of `torch._C._autocast_supported_devices()`, not just cpu/cuda), plus the shared `cache_enabled` flag and nesting counter -- following the `(label, getter, setter)`-tuple pattern already used by `torch/testing/_internal/common_utils.py`'s `_fp32_precision_flag_specs()`/`_snapshot_fp32_precision()`/`_restore_fp32_precision()` for the structurally identical problem of leaking backend flags between tests. `_autocast_state_specs()` is the single source of truth for what gets snapshotted and how to restore it; it's cached with `@functools.cache` since the device list is fixed at compile time. `tearDown`'s diagnostic on a detected leak lists each changed field's before/after value.

There is no direct Python getter for the nesting counter, only `autocast_increment_nesting()`/`autocast_decrement_nesting()`, so it's read via a no-net-effect increment+decrement pair (`_autocast_nesting()`) and restored by walking it back to the snapshotted depth with a delta-sized loop of increment/decrement calls (`_restore_autocast_nesting()`), which also calls `torch.clear_autocast_cache()` whenever a decrement brings the counter to zero -- mirroring `torch.autocast.__exit__` exactly, since skipping that clear leaves stale cast tensors reachable by later tests.

`tearDown` wraps `super().tearDown()` in a `try`/`finally` so the restore always runs even if `super().tearDown()` raises (its own leak detectors -- torch-function-mode-stack, fp32-precision flags, default dtype -- can themselves raise); the restore itself runs grad-mode, autocast, and `config.nested_graph_breaks` restoration as a plain sequential helper, since none of those three steps can realistically fail on their own (they are simple TLS/attribute writes restoring a value that was already valid when snapshotted).

The regression tests moved to a new dedicated file, `test/dynamo/test_test_case.py`, rather than living in `test/dynamo/test_ctx_manager.py`: they exercise `TestCase`'s own `setUp`/`tearDown` lifecycle directly and never compile anything, whereas every other test in `test_ctx_manager.py` compiles a function that uses a context manager, so they didn't fit that file's actual subject.

### Test Plan

`test/dynamo/test_test_case.py` adds 6 tests: `test_teardown_restores_autocast_{enabled,dtype,cache_enabled}` each leak one piece of autocast state through a real `torch._dynamo.test_case.TestCase` instance's `setUp`/`tearDown` cycle (instantiated directly with `methodName="runTest"`, not through the `unittest` runner) and check state is back to baseline afterward; `test_teardown_restores_autocast_nesting_after_increment` and `_after_decrement` cover both directions of the nesting-counter restore, the former also asserting `torch.clear_autocast_cache()` is called exactly once via `mock.patch.object(wraps=...)`; and `test_teardown_restores_autocast_after_super_teardown_error` forces a real (unmocked) `super().tearDown()` failure via a default-dtype mismatch and confirms both grad mode and autocast state are still restored despite it.

Verified via a temporary same-content overlay onto a second, fully built checkout (this worktree's own editable install does not resolve to itself), all 6 new tests individually, then the full suites:

```
python test/dynamo/test_test_case.py
python test/dynamo/test_ctx_manager.py
python test/dynamo/test_misc.py
```

`test_test_case.py`: 6/6 pass. `test_ctx_manager.py`: 141 tests, 0 failures (down from the pre-existing 147 by exactly the 6 tests moved out, confirming a clean move). `test_misc.py`: 895 tests, 0 failures.

```
python -m spin lint
```

Clean on the changed files (`spin lint`'s CLANGTIDY step cannot run without a build directory in a fresh worktree; verified separately via `lintrunner` scoped to the changed files and the relevant Python/test linters, also clean).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by aperson30** [comment]
@jansel Thanks for the review both were real. Pushed an update:

clear_autocast_cache() now called on nesting restore, mirroring __exit__.
tearDown restructured so every restore step is independent and a restore-time failure never masks the original exception.
Widened to "cpu" + "cuda", dropped the speculative guard, added test coverage for the exception-safety paths.
Note: test_autocast_object_guarded_by_value_not_identity fails in my env, but also fails identically on a clean upstream/main -- pre-existing, unrelated.

Left as-is (flagging, not fixing): whether this belongs in common_utils.py's generic leak detector instead, and device coverage beyond cpu/cuda as they are out of scope. Happy to take either further if you think it's worth it.

**Comment by rtimpe** [comment]
See the AI-generated review below.  I think it's worth handling all device types if we're going to do this (and also making it a bit more efficient - finding 5 below).

 

> **1. The autocast snapshot only covers cpu/cuda.** Autocast supports 10 device
>   types, and Dynamo's own guard already keys on all of them
>   (`torch/csrc/dynamo/guards.cpp`, `AutocastState::DEVICES`), as does the
>   AOTAutograd cache key. So a test that leaks e.g. `xpu` autocast state gets no
>   warning, no restore, and can trigger spurious recompiles in later tests in the
>   same process — exactly what this is meant to prevent. I confirmed the leak
>   survives `tearDown()` and that flipping `xpu` autocast does force a recompile.
> 
>   There's already a canonical list exposed as
>   `torch._C._autocast_supported_devices()` (used in
>   `_functorch/_aot_autograd/autograd_cache.py`), so this is a one-line swap for
>   the hardcoded `("cpu", "cuda")`. I measured the extra cost at ~15us per
>   snapshot, ~1.4s across the whole Dynamo suite. The `_autocast_state()` helper
>   in the test file needs the same change to stay in sync.
> 
>   **2. The error handling in `_restore_prior_test_state` looks like more than is
>   needed.** None of the three restores can realistically raise — they're all TLS
>   writes, and `config.nested_graph_breaks = ...` only raises for an unknown config
>   key. And when two exceptions do collide, Python's implicit chaining already
>   surfaces both; the `errors` list plus `log.error(..., exc_info=...)` is
>   reimplementing that to control which one is primary.
> 
>   The `try/finally` around `super().tearDown()` is clearly worth keeping — that's
>   a real fix, since the default-dtype and CUDA leak checks can throw and used to
>   skip the restore. But I think the rest collapses to a plain
>   `try: super().tearDown() / finally: self._restore_prior_test_state()`. I tried
>   that locally: 6 of the 8 new tests still pass, including
>   `test_teardown_restores_autocast_after_super_teardown_error` (the one that trips
>   a real, unmocked teardown failure). The two that fail are the two that only test
>   the aggregation machinery, and
>   `test_teardown_restore_error_not_suppressed_by_handled_exception` becomes
>   vacuous. Worth noting that those three tests are reachable only by mocking the
>   function under test to raise, which is usually a sign the branch isn't real.
> 
>   **3. The increment branch of `_restore_autocast_nesting` is untested.** The
>   nesting test only leaks upward, so the `for _ in range(-delta)` loop never runs.
>   A test that exits one more autocast context than it enters takes that path. It's
>   correct as written — just unguarded against regression.
> 
>   **4. `test_teardown_restores_autocast_nesting` assumes baseline nesting is 0.**
>   `clear_autocast_cache()` is only called when a decrement returns 0, so
>   `assert_called_once_with()` holds only if nothing outside the test has
>   incremented nesting. Under reordering, or after a test that leaks nesting
>   upward, it fails for an unrelated reason. An explicit assert on the precondition
>   would make that obvious.
> 
>   **5. Minor:** `_autocast_state_specs()` rebuilds its `partial`s on every call,
>   and it's called 3+ times per test (more once the device list grows). The list is
>   static, so `@functools.cache` would do.

**Comment by aperson30** [comment]
@rtimpe @jansel  Thanks for the review. Pushed another update addressing the feedback:

Simplified _restore_prior_test_state back to a plain sequential restore (grad mode → autocast → config.nested_graph_breaks) instead of the error-collection machinery -- none of those three steps can realistically fail on their own, so the extra complexity wasn't earning its keep.
Moved the regression tests out of test_ctx_manager.py into a new test/dynamo/test_test_case.py, since they test TestCase's own lifecycle directly and never compile anything.
tearDown's warning on a detected leak now includes before/after values per field, not just the field names.
Tests, full suites, and lint all still green.

**Review by rtimpe — APPROVED** [review]

**Comment by aperson30** [comment]
@rtimpe @jansel  sorry for the inconvenience but these failures seem to be unrelated to this pr could one of you do a merge-i  when you get the chance. Thanks.
